File: djangoBackend2/my_app/services/PostInfoFetchingService.py
from ..models import UserFollowing, UserBlocking

import requests

from django.db.models import Q

def check_if_auth_user_is_an_author_of_post(auth_user_id, overall_post_id):
    try:
        response = requests.get(
            f'http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost/{overall_post_id}'
        )

        if response.status_code == 404:
            return [
                'There doesn\'t currently exist a post with the overallPostId that you provided.',
                'NOT_FOUND'
            ]

        if not response.ok:
            return [
                'The expressJSBackend1 server had trouble getting the authors of the post.',
                'BAD_GATEWAY'
            ]

        response_data = response.json()
        authors_of_post = response_data.get('authorsOfPost')

        return auth_user_id in authors_of_post
    
    except:
        return [
            'There was trouble connecting to the expressJSBackend1 server to get the authors of the post.',
            'BAD_GATEWAY'
        ]


def check_if_auth_user_has_access_to_post(auth_user_id, overall_post_id):
    authors_of_post = []
    set_of_authors_of_post = None
    is_encrypted = False

    try:
        response = requests.get(
            f'http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost/{overall_post_id}'
        )

        if response.status_code == 404:
            return [
                'There doesn\'t currently exist a post with the overallPostId that you provided.',
                'NOT_FOUND'
            ]

        if not response.ok:
            return [
                'The expressJSBackend1 server had trouble getting the authors of the post.',
                'BAD_GATEWAY'
            ]

        response_data = response.json()
        authors_of_post = response_data.get('authorsOfPost')
        set_of_authors_of_post = set(authors_of_post)
        is_encrypted = response_data.get('isEncrypted')

        if auth_user_id in authors_of_post:
            return True

    except:
        return [
            'There was trouble connecting to the expressJSBackend1 server to get the authors of the post.',
            'BAD_GATEWAY'
        ]

    if is_encrypted:
        try:
            user_follows_at_least_one_post_author = UserFollowing.objects.filter(
                follower=auth_user_id,
                followed__in=set_of_authors_of_post 
            ).exists()


            if not user_follows_at_least_one_post_author:
                return [
                    f'User {auth_user_id} does not follow at-least one of the authors of post {overall_post_id} and hence
                    does not have the authority to access/modify this encrypted post\'s data',
                    'UNAUTHORIZED'
                ]

        except:
            return [
                f'There was trouble connecting checking whether or not user {auth_user_id} follows at-least one of the authors of
                the post',
                'BAD_GATEWAY'
            ]

    else:
        try:
            each_author_of_post_is_in_blockings_of_auth_user = (UserBlocking.objects
                .filter(
                    Q(blocker=auth_user_id, blocked__in=set_of_authors_of_post) |
                    Q(blocker__in=set_of_authors_of_post, blocked=auth_user_id)
                )
                .count()
            ) == len(set_of_authors_of_post)

            if each_author_of_post_is_in_blockings_of_auth_user:
                return [
                    'There doesn\'t currently exist a post with the overallPostId that you provided.',
                    'NOT_FOUND'
                ]

        except:
            return [
                f'There was trouble checking whether or not each of the authors of this unencrypted post either block you or are
                blocked by you.',
                'BAD_GATEWAY'
            ]

    return True

File: laravelBackend1/app/Services/PostInfoFetchingService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Http;


class PostInfoFetchingService {


    public function getPostEncryptionStatusIfUserHasAccessToPost(int $authUserId, string $overallPostId) {  
        $authorsOfPost = [];
        $isEncrypted = false;

        try {
            $response = Http::get(
                "http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost/$overallPostId"
            );

            
            if ($response->failed()) {
                if ($response->status() == 404) {
                    return [
                        "There doesn't currently exist a post with the overallPostId that you provided.",
                        'NOT_FOUND'
                    ];
                }

                return [
                    "The expressJSBackend1 server had trouble getting the authors and encryption-status of the
                    post.",
                    'BAD_GATEWAY'
                ];
            }

            $stringifiedResponseData = $response->body();
            $parsedResponseData = json_decode($stringifiedResponseData, true);

            $authorsOfPost = $parsedResponseData["authorsOfPost"];
            $isEncrypted = $parsedResponseData["isEncrypted"];
            
            if (in_array($authUserId, $authorsOfPost)) {
                return $isEncrypted;
            }
        }
        catch (\Exception) {
            return [
                "There was trouble connecting to the expressJSBackend1 server to get the authors and encryption-
                status of the post.",
                "BAD_GATEWAY"
            ]; 
        }

        if ($isEncrypted) {
            if ($authUserId == -1) {
               return [
                    "As an anonymous guest, you do not have access to this private-post or any of the encrypted
                    data associated with it.",
                    "UNAUTHORIZED"
               ];  
            } 

            try {
                $response1 = Http::withHeaders([
                    'Content-Type' => 'application/json',
                ])->post(
                    "http://34.111.89.101/api/Home-Page/djangoBackend2/graphql",
                    [
                        'query' => 'query ($authUserId: Int!, $userIds: [Int!]!) {
                            checkIfUserFollowsAtLeastOneInList(authUserId: $authUserId, userIds: $userIds)
                        }',
                        'variables' => [
                            'authUserId' => $authUserId,
                            'userIds' => $authorsOfPost
                        ]
                    ]
                );
                
                if ($response1->failed()) {
                    return [
                        "The djangoBackend2 server had trouble verifying whether or not you follow at-least one of the
                        authors of this private-post.",
                        "BAD_GATEWAY"
                    ]; 
                }

                $parsedResponse1Data = $response1->json();
                $userFollowsAtLeastOneAuthor = (bool) $parsedResponse1Data['data']['checkIfUserFollowsAtLeastOneInList'];

                if (!$userFollowsAtLeastOneAuthor) {
                    return [
                        "You do not have access to any of the encrypted-data of this post since you do not follow
                        at-least one of its authors.",
                        "UNAUTHORIZED"
                    ]; 
                }
            }
            catch (\Exception) {
                return [
                    "There was trouble connecting to the djangoBackend2 server to verify whether or not you follow at-least 
                    one of the authors of this private-post.",
                    "BAD_GATEWAY"
                ]; 
            }  
        }
        else {
            if ($authUserId !== -1) {
                try
                {
                    $response2 = Http::withHeaders([
                        'Content-Type' => 'application/json'
                    ])->post("http://34.111.89.101/api/Home-Page/djangoBackend2/isEachUserInListInTheBlockingsOfAuthUser
                    /$authUserId", [
                        'user_ids' => $authorsOfPost
                    ]);

                    if ($response2->failed()) {
                        return [
                            "The djangoBackend2 server had trouble checking whether or not
                            each of the authors of this post either block you or are blocked by you.",
                            "BAD_GATEWAY"
                        ]; 
                    }

                    $stringifiedResponse2Data = $response2->body();
                    $eachPostAuthorIsInAuthUserBlockings = (bool) $stringifiedResponse2Data;
                    if ($eachPostAuthorIsInAuthUserBlockings)
                    {
                        return [
                            "You are trying to access the data of a post that does not exist.",
                            "NOT_FOUND"
                        ];
                    }
                }
                catch (\Exception) {
                    return [
                        "There was trouble connecting to the djangoBackend2 server to check whether or not
                        each of the authors of this unencrypted post either block you or are blocked by you.",
                        "BAD_GATEWAY"
                    ];
                }
            }
        }

        return $isEncrypted;
    }

    
    public function ifUserIsAnAuthorOfPostGetEncryptionStatus(int $authUserId, string $overallPostId) {
        try {
            $response = Http::get(
                "http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost/$overallPostId"
            );

            if ($response->failed()) {
                if ($response->status() == 404) {
                    return [
                        "There doesn't currently exist a post with the overallPostId that you provided.",
                        'NOT_FOUND'
                    ];
                }

                return [
                    "The expressJSBackend1 server had trouble getting the authors and encryption-status of the
                    post.",
                    'BAD_GATEWAY'
                ];
            }

            $stringifiedResponseData = $response->body();
            $parsedResponseData = json_decode($stringifiedResponseData, true);

            $authorsOfPost = $parsedResponseData["authorsOfPost"];
            $isEncrypted = $parsedResponseData["isEncrypted"];
            
            if (in_array($authUserId, $authorsOfPost)) {
                return $isEncrypted;
            }
            return [
                "You are not one of the authors of the post with this overallPostId-> $overallPostId",
                "UNAUTHORIZED"
            ];
        }
        catch (\Exception) {
            return [
                "There was trouble connecting to the expressJSBackend1 server to get the authors and encryption-
                status of the post.",
                "BAD_GATEWAY"
            ]; 
        }
    }
}